File: 0xcf.py
from collections import deque

cases = int(input())

for case in range(cases):
	n = input()
	buildings = [int(v) for v in input().split(" ")]
	
	result = []
	stack = deque([buildings[-1]])

	for building in buildings[::-1]:
			
		while len(stack) > 0 and building >= stack[-1]:
			stack.pop()

		if len(stack) > 0 and building < stack[-1]:
			result.insert(0, stack[-1])
		if len(stack) == 0:
			result.insert(0, -1)

		stack.append(building)

	str_result = ""
	for building in result:
		str_result += str(building) + " "

	print(f"Case #{case+1}: {str_result}")

File: 0xcf.java
import java.io.*;
import java.util.Stack;

public class Main {
	public static void main(String args[]) throws IOException {
		BufferedReader bfr = new BufferedReader(new InputStreamReader(System.in));
		int cases = Integer.parseInt(bfr.readLine());
		BufferedWriter bfw = new BufferedWriter(new OutputStreamWriter(System.out));
		for (int c = 1; c <= cases; c++) {
			int n = Integer.parseInt(bfr.readLine());
			
			int buildings[] = new int[n];
			int result[] = new int[n];
			String buildingString[] = bfr.readLine().split(" ");
			Stack<Integer> stack = new Stack<Integer>();

			for (int i = 0; i < n; i++) {
				buildings[i] = Integer.parseInt(buildingString[i]);
			}

			for (int i = n - 1; i >= 0; i--) {
				int building = buildings[i];
				while (!stack.empty() && building >= stack.peek()) {
					stack.pop();
				}
				
				if(!stack.empty() && building < stack.peek()) {
					result[i] = stack.peek();
				}

				if(stack.empty()) {
					result[i] = -1;
				}

				stack.push(building);
			}

			StringBuilder sb = new StringBuilder();
			
			for (int building : result){
				sb.append(building).append(" ");
			}

			bfw.write("Case #" + c + ": " + sb.toString() + "\n");
		}
		bfw.flush();
	}
}
